crates/iota-data-ingestion-core/src/worker_pool.rs

```diff
@@ -82,78 +205,35 @@ impl<W: Worker + 'static> WorkerPool<W> {
-                            if let Some(watermark) =  progress_watermark {
-                                executor_status_update = Some(watermark + 1);
-                            }
-                            current_checkpoint_number += 1;
-                        }
-                        if let Some(update) = executor_status_update {
-                            if executor_progress_sender
-                                .send((self.task_name.clone(), update))
-                                .await.is_err() {
-                                    // The executor progress channel closing is a sign we need to
-                                    // exit this loop.
-                                    break;
-                                }
-                        }
-                    }
-                    while !checkpoints.is_empty() && !idle.is_empty() {
-                        let checkpoint = checkpoints.pop_front().unwrap();
-                        let worker_id = idle.pop_first().unwrap();
-                        if workers[worker_id].0.send(checkpoint).await.is_err() {
-                            // The worker channel closing is a sign we need to exit this loop.
-                            break;
-                        }
-                    }
-                }
-                Some(checkpoint) = checkpoint_receiver.recv() => {
-                    let sequence_number = checkpoint.checkpoint_summary.sequence_number;
-                    if sequence_number < current_checkpoint_number {
-                        continue;
-                    }
-                    self.worker.preprocess_hook(checkpoint.clone()).expect("failed to preprocess task");
-                    if idle.is_empty() {
-                        checkpoints.push_back(checkpoint);
-                    } else {
-                        let worker_id = idle.pop_first().unwrap();
-                        if workers[worker_id].0.send(checkpoint).await.is_err() {
-                            // The worker channel closing is a sign we need to exit this loop.
-                            break;
-                        };
-                    }
-                }
-            }
-        }
-
-        // Clean up code for graceful termination
-
-        // Notify the exit handles of all workers to terminate
-        drop(workers);
+        (worker_senders, workers_join_handles)
+    }
 
-        // Wait for all workers to finish
-        for join_handle in join_handles {
-            join_handle.await.expect("worker thread panicked");
+    /// Start the child workers graceful shutdown
+    async fn workers_graceful_shutdown(
+        &self,
+        workers_join_handles: Vec<JoinHandle<()>>,
+        executor_progress_sender: mpsc::Sender<WorkerPoolStatus>,
+    ) {
+        for worker in workers_join_handles {
+            // Await the Worker actor completion
+            worker.await.expect("worker thread panicked");
```

Comment:
are you sure you want to panic? This will prevent the remaining workers from shutting down gracefully, no? Maybe it would be better to error log and await the other workers?

Reply:
You're right, thank you catching this! [Fixed](https://github.com/iotaledger/iota/pull/4648/commits/2eda721b05e84526cb29cc4bb968b326dbb206da#diff-dfef1eeaea282ac0a41b60863611b2244808890824a073bf2264ca6fc775a624R233-R235)